### PR TITLE
feat(pubsub): faster subscription shutdowns

### LINKED ISSUE #5026: Speed up Pub/Sub shutdown
Currently the shutdown can block for a long time (about 90s) waiting for a response from the last `AsyncPull()`.

### PR DESCRIPTION
**Pull request by coryan**
Finally use the `AsyncStreamingPull()` RPC to reduce the shutdown time
for subscription sessions. I have anecdotal evidence that this also
improves the overall throughput.

Part of the work for #5192 and fixes #5026

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/5198)

### CODE REVIEW

File: google/cloud/pubsub/BUILD [hunk 1]
```diff
@@ -40,6 +40,7 @@ cc_library(
         "//google/cloud:google_cloud_cpp_common",
         "//google/cloud:google_cloud_cpp_grpc_utils",
         "@com_google_googleapis//google/pubsub/v1:pubsub_cc_grpc",
+        "@com_google_googletest//:gtest_main",
```
Comment (devbww) [inline google/cloud/pubsub/BUILD:43]: It seems strange for a library to depend on `gtest_main`.  Should this just be `gtest`?

File: google/cloud/pubsub/subscriber_connection.cc [hunk 2]
```diff
@@ -55,21 +56,30 @@ class SubscriberConnectionImpl : public pubsub::SubscriberConnection {
...
   ~SubscriberConnectionImpl() override = default;
 
   future<Status> Subscribe(SubscribeParams p) override {
-    return CreateSubscriptionSession(stub_, background_->cq(), std::move(p),
-                                     retry_policy_->clone(),
-                                     backoff_policy_->clone());
+    auto client_id = [this] {
+      std::lock_guard<std::mutex> lk(mu_);
+      auto constexpr kLength = 32;
+      auto constexpr kChars = "abcdefghijlkmnopqrstuvwxyz0123456789";
```
Comment (devbww) [inline google/cloud/pubsub/subscriber_connection.cc:68]: s/jlk/jkl/?

File: google/cloud/pubsub/internal/subscription_session_test.cc [hunk 3]
```diff
@@ -614,54 +350,14 @@ TEST(SubscriptionSessionTest, ShutdownWaitsEarlyAcks) {
...
+  EXPECT_CALL(*mock, AsyncStreamingPull)
+      .Times(AtLeast(1))
+      .WillRepeatedly(FakeAsyncStreamingPull);
+
   auto constexpr kMessageCount = 16;
-  std::mutex generate_mu;
-  int generate_count = 0;
 
   // These lambdas are used to implement the mocks. They become satisfied using
   // a timer, which better simulates the behavior in production.
```
Comment (devbww) [inline google/cloud/pubsub/internal/subscription_session_test.cc:360]: Orphaned comment?

Comment (coryan) [reply google/cloud/pubsub/internal/subscription_session_test.cc]: Fixed.

File: google/cloud/pubsub/subscriber_connection.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/pubsub/subscriber_connection.cc]: Fixed

File: google/cloud/pubsub/BUILD (hunk 1 above)
Comment (coryan) [reply google/cloud/pubsub/BUILD]: Fixed

**Review by devbww — APPROVED** [review]

**Pull request merged**